### ISSUE TITLE
Reusable Observables

### ISSUE THREAD

**Issue by gNechitailo** [marblejs/marble#181]
**Is your feature request related to a problem? Please describe.**
I have read the description to this project and I like the idea of functional programming related to Web development. The point I didn't get is that you call HTTP single-event based. HTTP has a set of declared endpoints with a declared processing logic.
I researched a bit about RxJS and found out that creating observable every time you want to process an event is WAY MORE expensive than using the same observable and pushing new events to it.
I have also used a debugger a bit with Marble.js and saw that it really creates a new observable for each request. The nature of EffectFactory about that.

**Describe the solution you'd like**
I think if marble used effectFactory callback upon application startup to create a processing pipe Once instead of calling it on each request, it would be much more performant.
Maybe you'll need some extra RxJS operators to perform routing, but I can't tell for sure (didn't dive deep to marble.js sources).

**Additional context**
Maybe there are there fundamental problems about it?
If not, I'd like to make some help with this subject (maybe make a version for this).

**Comment by JozefFlakus** [comment]
Hi @gNechitailo,
Thanks for the interest! Good spot. The way how the routing in `@marblejs/core` was build had two  approaches:
1. stream based routing (without static routing table - dynamic routing) **Marble 0.5**
2. router based (streams like callbacks but with static routing build on startup) **Marble 1.0**

The diagram below shows how how the server reacted to high load of incoming request (req/sec) depending of endpoints total number. As you can see it is quite performant right now and it was the main aim when releasing stable version 1.

![routing-chart-2](https://user-images.githubusercontent.com/7793430/69572194-e8faa080-0fc3-11ea-9fff-70bc9fc05c39.png)

![chart-1](https://user-images.githubusercontent.com/7793430/69572465-7c33d600-0fc4-11ea-9b44-b537574feb8d.png)

The mechanism that you describing would be an evolution of the current solution. Basically on the server startup the built routing table should be reflected as a stream with proper operators acting as filters and matchers for method and routing paths. This is indeed possible but we have to be very careful, especially we have to be sure the stream will be always active - it cannot be closed due to errors.

The significant benefit of this would be in Context dependency injection, eg. you will be able to resolve the declared dependencies inside the callback (instead of stream), so the dips will be always injected after bootraping.

The idea of this always running steam I'm already implementing in upcoming `@marblejs/messaging` module in version 3 but recently I'm thinking about such an improvement quite a lot.

tl;dr I have to crate a POC of this solution if it is possible to reflect the collected routing table as a static stream.

**Comment by gNechitailo** [comment]
Wow! Thank you for the detailed answer!
Keep up the good work!

**Comment by JozefFlakus** [comment]
In general, the proposed approach has to be testes, it can turn out that in the end it will be slower than the current implementation, but the memory overhead will be lower (which is not so bad right now).

**Comment by JozefFlakus** [comment]
@gNechitailo checkout PR number #188 - I think it should be interesting fo you.

**Comment by gNechitailo** [comment]
Wow! Amazing! This is my favorite library now :) 
Tried to start this work myself, but you were much faster (it wasn't that easy).

**Comment by JozefFlakus** [comment]
Glad to hear that! I believe that it can be optimized even more, but right now it is a good base for future improvements.

**Comment by JozefFlakus** [comment]
@gNechitailo It turned out that the implementation previously introduced was buggy so I had to revert to older implementation. If you would like to be interested in problem-solving. Any help highly appreciated! :)

**Comment by gNechitailo** [comment]
Okay, I will try to take a look when I have spare time!
